app/src/main/java/com/xmartlabs/gong/ui/screens/signin/SignInScreenViewModel.kt

```diff
@@ -1,51 +1,37 @@
 package com.xmartlabs.gong.ui.screens.signin
 
-import androidx.annotation.MainThread
-import androidx.lifecycle.LiveData
-import androidx.lifecycle.MutableLiveData
-import androidx.lifecycle.ViewModel
-import androidx.lifecycle.asLiveData
-import androidx.lifecycle.switchMap
-import com.xmartlabs.gong.data.model.User
+import androidx.lifecycle.viewModelScope
 import com.xmartlabs.gong.device.common.ProcessState
 import com.xmartlabs.gong.domain.usecase.SignInUseCase
-import com.xmartlabs.gong.domain.usecase.TimeTrackerUseCase
-import java.util.Date
-import kotlin.time.Duration
-import kotlin.time.ExperimentalTime
+import com.xmartlabs.gong.ui.common.BaseViewModel
 
 /**
  * Created by mirland on 25/04/20.
  */
 class SignInScreenViewModel(
     private val signInUseCase: SignInUseCase,
-    private val timeTrackerUseCase: TimeTrackerUseCase,
-) : ViewModel() {
-
-  private val signInMutableLiveData = MutableLiveData<SignInUseCase.Params>()
-  @OptIn(ExperimentalTime::class)
-  val viewModelTime: LiveData<Result<Duration>> = timeTrackerUseCase.invoke(TimeTrackerUseCase.Params(Date()))
-      .asLiveData()
-  val signIn: LiveData<ProcessState<User>> = signInMutableLiveData
-      .switchMap { params -> signInUseCase.invokeAsLiveData(params) }
-
-  private val userMutableLiveData = MutableLiveData<String>()
-  val userLiveData = userMutableLiveData
-
-  var passwordMutableLiveData = MutableLiveData<String>()
-  var passwordLiveData = passwordMutableLiveData
-
-  @MainThread
-  fun signIn() {
-    signInMutableLiveData.value =
-        SignInUseCase.Params(userLiveData.value ?: "", passwordLiveData.value ?: "")
-  }
-
-  fun updateUser(user: String) {
-    userMutableLiveData.value = user
+) : BaseViewModel<
+    SignInUiAction,
+    SignInViewModelEvent,
+    SignInViewState
+    >(SignInViewState()) {
+
+  override suspend fun processAction(action: SignInUiAction) = when (action) {
+    SignInUiAction.SignIn -> signIn()
+    is SignInUiAction.ChangePassword -> setState { copy(password = action.password) }
+    is SignInUiAction.ChangeUserName -> setState { copy(userName = action.userName) }
   }
 
-  fun updatePassword(password: String) {
-    passwordMutableLiveData.value = password
+  private suspend fun signIn() {
+    viewModelScope.launchWithState { state ->
+      signInUseCase.invokeAsFlow(SignInUseCase.Params(state.userName, state.password))
+          .watchProcessState { state ->
+            setState { copy(isLoading = state == ProcessState.Loading) }
```

Comment:
maybe an extension to ask if state is loading?

Reply:
how? I didn't understand, the state has an `isLoading` property

Reply:
yeah but just `state.isLoading()` instead of `state == ProcessState.Loading`

Reply:
oh I got it, yes but in my opinion, if you do that you should have 3 extensions, isLoading, isSuccess, isFailure, and I think they are not needed, you will lose the magic of sealed classes, you are like hiding it and I think it's important, for example:
```
val foo = when ){
 state is Success || state is Failure -> something
 state.isLoading() -> TODO()
}
```
this will not compile, and if you replace ` state.isLoading() ` with `state == ProcessState.Loading` will compile, and we should use it in the same way in all cases, that's why I don't like it. However, I can change it, it's not a big change and I can do it. what do you think?

Reply:
I could find something interesting using contracts